service/turn.rest/src/main/java/org/kaazing/gateway/service/turn/rest/TurnRestServiceHandler.java

```diff
@@ -0,0 +1,110 @@
+import org.kaazing.mina.core.buffer.IoBufferAllocatorEx;
+import org.kaazing.mina.core.buffer.IoBufferEx;
+
+class TurnRestServiceHandler extends IoHandlerAdapter<HttpAcceptSession> {
+    
+    private ServiceProperties options;
+    private KeyStore keystore;
+    private TurnRestCredentialsGenerator credentialGenerator;
+    private String uris;
+
+    TurnRestServiceHandler(ServiceProperties options, KeyStore keystore, TurnRestCredentialsGenerator credentialGenerator, String uris) {
+        this.options = options;
+        this.keystore = keystore;
+        this.credentialGenerator = credentialGenerator;
+        this.uris = uris;
+    }
+
+    @Override
+    protected void doSessionOpened(HttpAcceptSession session) throws Exception {        
+        HttpMethod method = session.getMethod();
+        String service = session.getParameter("service");
+      
+        String ttl = options.get("credentials.ttl");
+        Certificate alias = keystore.getCertificate(options.get("key.alias"));
+        char separator = options.get("username.separator").charAt(0);
+         
+        if (method != HttpMethod.GET) {
+            session.setStatus(HttpStatus.CLIENT_METHOD_NOT_ALLOWED);
+            session.close(false);
+            throw new IllegalArgumentException("HTTP method not allowed: " + method);        
+        } else if (!service.equals("turn")) {
+            session.setStatus(HttpStatus.CLIENT_BAD_REQUEST);
+            session.close(false);
+            throw new IllegalArgumentException("Unsupported/invalid service: " + service);
+        }
+        
+        session.setVersion(HttpVersion.HTTP_1_1);
+        session.setWriteHeader(HttpHeaders.HEADER_CONTENT_TYPE, "application/json");
+ 
+        Subject subject = session.getSubject();
+        
+        if (ttl == null) {
+            ttl = session.getParameter("Max-Age");
+        }
+       
+        TurnRestCredentials credentials = null;
+        String username = null;
+        char[] password = null;
+        if (credentialGenerator != null) {
+            credentialGenerator.setCredentialsTTL(ttl);
+            credentialGenerator.setKeyAlias(alias);
+            credentialGenerator.setUsernameSeparator(separator);
+            credentials = credentialGenerator.generate(subject);
+            username = credentials.getUsername();
+            password = credentials.getPassword();
+        }
+     
+        String response = TurnRestJSONResponse.createResponse(username, password, ttl, this.uris); 
+        CharSequence responseChars = response;
+        
+        if (password != null) {
+            Arrays.fill(password, '0');
+        }
+        
+        // get io buffer for file
+        IoBufferAllocatorEx<?> allocator = session.getBufferAllocator();
+        IoBufferEx out = allocator.wrap(allocator.allocate(response.length())).setAutoExpander(allocator);
+        out.put(response.getBytes());
+        out.putString(responseChars, StandardCharsets.UTF_8.newEncoder());
```

Comment:
Why is there a need to do a `putString` after the initial `put`? This will duplicate the response.

Additionally a `out.flip()` should be added here, otherwise it won't work